pkg/webhooks/validation/controller/controller.go

```diff
@@ -0,0 +1,530 @@
+		return err
+	}
+	scope.Info("Successfully deleted validatingwebhookconfiguration")
+	return nil
+}
+
+func (c *Controller) updateValidatingWebhookConfiguration(desired *kubeApiAdmission.ValidatingWebhookConfiguration) error {
+	current, err := c.sharedInformers.Admissionregistration().V1beta1().
+		ValidatingWebhookConfigurations().Lister().Get(c.o.WebhookConfigName)
+
+	if kubeErrors.IsNotFound(err) {
+		_, err := c.o.Client.AdmissionregistrationV1beta1().
+			ValidatingWebhookConfigurations().Create(desired)
+		if err != nil {
+			scope.Errorf("Failed to create webhook config: %v", err)
+			reportValidationConfigUpdateError(kubeErrors.ReasonForError(err))
+			return err
+		}
+		scope.Info("Successfully created webhook config")
+		reportValidationConfigUpdate()
+		return nil
+	}
+
+	updated := current.DeepCopyObject().(*kubeApiAdmission.ValidatingWebhookConfiguration)
+	updated.Webhooks = desired.Webhooks
+	updated.OwnerReferences = desired.OwnerReferences
+
+	if !reflect.DeepEqual(updated, current) {
+		_, err := c.o.Client.AdmissionregistrationV1beta1().
+			ValidatingWebhookConfigurations().Update(updated)
+		if err != nil {
+			scope.Errorf("Failed to update webhook config: %v", err)
+			reportValidationConfigUpdateError(kubeErrors.ReasonForError(err))
+			return err
+		}
+	}
+	scope.Info("Successfully created webhook config")
+	reportValidationConfigUpdate()
+	return nil
+}
+
+type configError struct {
+	err    error
+	reason string
+}
+
+func (e configError) Error() string {
+	return e.err.Error()
+}
+
+func (e configError) Reason() string {
+	return e.reason
+}
+
+func (c *Controller) buildValidatingWebhookConfiguration() (*kubeApiAdmission.ValidatingWebhookConfiguration, error) {
+	webhook, err := c.readFile(c.o.WebhookConfigPath)
+	if err != nil {
+		return nil, &configError{err, "could not read configuration file"}
+	}
+	caBundle, err := c.readFile(c.o.CAPath)
+	if err != nil {
+		return nil, &configError{err, "could not read caBundle file"}
+	}
+	return buildValidatingWebhookConfiguration(caBundle, webhook, c.ownerRefs)
+}
+
+func buildValidatingWebhookConfiguration(
+	caBundle, webhook []byte,
+	ownerRefs []kubeApiMeta.OwnerReference,
+) (*kubeApiAdmission.ValidatingWebhookConfiguration, error) {
+	config, err := decodeValidatingConfig(webhook)
+	if err != nil {
+		return nil, &configError{err, "could not decode webhook configurationd"}
+	}
+	if err := verifyCABundle(caBundle); err != nil {
+		return nil, &configError{err, "could not verify caBundle"}
+	}
+	// update runtime fields
+	config.OwnerReferences = ownerRefs
+	for i := range config.Webhooks {
+		config.Webhooks[i].ClientConfig.CABundle = caBundle
+	}
+
+	return config, nil
+}
+
+func decodeValidatingConfig(encoded []byte) (*kubeApiAdmission.ValidatingWebhookConfiguration, error) {
+	var config kubeApiAdmission.ValidatingWebhookConfiguration
+	if err := runtime.DecodeInto(scheme.Codecs.UniversalDecoder(), encoded, &config); err != nil {
```

Comment:
Do we need the whole kubectl dependency just for this? how did we do it before?

Reply:
We don't need it after all. The decoder is suppose to handle defaulting to reduce unnecessary patching, but the configuration type doesn't have a registered defaulter. We'll need to handle this ourselves as we've done previously.